### PR TITLE
Auth plugin caching sha2 password

### PR DESCRIPTION
**Pull request by wenjoy**
It's related about #16 

Currently it works for both mysql 8.0 and mysql 5.6 based on my test.
But it just works. There is still work left to completely support mysql8.0. In further, we need to:
- [ ] support negotiation of auth method  if mismatched
- [ ] used configured public key
and others I doesn't figure out now.

Anyway, please help to check this PR and give me feedbacks. Thanks!

### CODE REVIEW

File: src/connection.ts [hunk 1]
```diff
@@ -40,7 +46,7 @@ export class Connection {
   constructor(readonly client: Client) {}
 
   private async _connect() {
-    const { hostname, port = 3306 } = this.client.config;
+    const { hostname, port = 3306, username, password } = this.client.config;
```
Comment (manyuanrong) [inline src/connection.ts:49]: `username = ""` should be better

**Comment by manyuanrong** [comment]
For the feature of configuring publicKey, it may be good to do it in the new PR, because testing it requires more changes to CI and test cases

File: src/connection.ts (hunk 1 above)
Comment (wenjoy) [reply src/connection.ts]: Fixed, thx.

**Comment by wenjoy** [comment]
> For the feature of configuring publicKey, it may be good to do it in the new PR, because testing it requires more changes to CI and test cases

Agreed, should we also add these plans to our TODO.

**Comment by manyuanrong** [comment]
> Agree, should we also add these plans to our TODO.

Of course this is the best

**Comment by invisal** [comment]
Works on mysql8 with caching sha2 enabled on my computer.

**Comment by robinbertram** [comment]
This PR also solved the "Got packets out of order" problem I had. Great work!

**Comment by webdev3301** [comment]
After reading all of these questions regarding sha2 password it is still unclear how to implement it.

Can anyone help with a working example? I've been waiting on this for over an year.

**Comment by webdev3301** [comment]
"It works on mysql8 with caching sha2 enabled on my computer."

How? Where? How's this working? Where do I find it working?

**Pull request merged**